File: frontend/src/schemas/execution.schema.ts
import { z } from 'zod'
import { ExecutionSideEnum, ExecutionSourceEnum } from '@/enums'

export const ExecutionSchema = z.object({
  id: z.string().uuid(),
  tradeId: z.string().uuid(),
  scalePlanId: z.string().uuid(),
  price: z.number().min(1, 'Price is required'),
  qty: z.number().min(1, 'Quantity is required'),
  commission: z.number().min(1, 'Cost is required'),
  notes: z.string().optional(),
  executedAt: z.date(),
  side: ExecutionSideEnum,
  source: ExecutionSourceEnum,
})

export const ExecutionCreateSchema = ExecutionSchema.omit({
  id: true,
  executedAt: true,
})

export const ExecutionUpdateSchema = ExecutionCreateSchema.partial()

File: backend/domain/execution/execution_schema.py
from datetime import datetime
from typing import Optional

from core.base_schema import BaseSchema
from database.models import Side, ExecSource


class ExecutionBase(BaseSchema):
    notes: Optional[str] = None
    commission: Optional[float] = None
    side: Optional[Side] = None
    source: Optional[ExecSource] = None


class ExecutionCreate(ExecutionBase):
    qty: int
    price: float
    live_trade_id: str
    scale_plan_id: str


class ExecutionRead(ExecutionCreate):
    id: str
    executed_at: datetime


class ExecutionUpdate(ExecutionBase):
    qty: Optional[int] = None
    price: Optional[float] = None
